### PR TITLE
std: Add a gzip decoder

### PR DESCRIPTION
**Pull request by LemonBoy**
There's no CRC16 in the stdlib :(

### CODE REVIEW

**Comment by andrewrk** [comment]
Nice work. I wonder what it would look like to try to do it without heap allocation.

This is progress towards package manager :muscle:

File: lib/std/compress/gzip.zig [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+const deflate = std.compress.deflate;
+
+// Flags for the FLG field in the header
+const FTEXT = 1 << 0;
+const FHCRC = 1 << 1;
+const FEXTRA = 1 << 2;
+const FNAME = 1 << 3;
+const FCOMMENT = 1 << 4;
+
+pub fn GzipStream(comptime ReaderType: type) type {
```
Comment (daurnimator) [inline lib/std/compress/gzip.zig:23]: Should this be `GzipStreamReader` with the expectation of a `GzipStreamWriter` one day being added?

Comment (LemonBoy) [reply lib/std/compress/gzip.zig:23]: I had planned to feed either a reader or a writer to the basic Stream type and have some comptime logic to detect the direction. But splitting the reader/writer is also fine (and possibly simpler)

**Comment by asafyish** [comment]
> Nice work. I wonder what it would look like to try to do it without heap allocation.
> 
> This is progress towards package manager 💪

You might want to consider zstd when the time comes. It has better compression ratio and over the long run, I think we should reduce network traffic as much as possible.

**Pull request merged**